### ISSUE TITLE
Use `JOIN`s in more queries

### ISSUE THREAD

**Issue by DakotaLarson** [prisma/prisma#5184]
## Problem

Currently, there is no way to join tables together. Queries that include relations only include the relational data by using separate queries.

## Suggested solution

I am hoping this can be implemented without a breaking API change. I would like to see a single query executed that joins multiple tables when necessary rather than a set of queries, each referencing a related table.

## Alternatives

The only alternative I can see is to use raw queries.

**Comment by DakotaLarson** [comment]
This is interesting: https://github.com/prisma/prisma/issues/4997#issuecomment-612429050. With this in mind, it is worth noting the main issue is not being able to only retrieve rows where the relational data is not null or empty (e.g. With the query above, only returning data where `otherTable` entries exist is not possible).

**Comment by sorenbs** [comment]
Does relation filters as described here solve your use case? https://www.prisma.io/docs/concepts/components/prisma-client/filtering/#filter-on-related-records

**Comment by DakotaLarson** [comment]
@sorenbs Yes! I'm not sure what long term plans are for this "optimization", but perhaps some more clarity in the docs about why Prisma uses multiple queries would create a better user experience so fewer issues are raised about this compromise. Long term, perhaps some way to opt in to using joins in a query would be helpful (as opposed to the library making a decision based on table size or schema).

**Comment by pantharshit00** [comment]
@DakotaLarson There is a talk about internal which goes over this: https://www.youtube.com/watch?v=RAoXdyI_PH4

You can refer to that for now.

**Comment by achadee** [comment]
Any news on this? I can see prisma breaking with high volume if it doesnt support joins

**Comment by SirMishaa** [comment]
I confess I do not understand how this is possible. This seems to me to be an essential feature for an ORM, especially nowadays where there are often many relationships. 

It simply makes Prisma unusable for heavy usage, especially with a lot of joins.

**Comment by janpio** [comment]
Please open an issue if something really is unusable or breaks in any way so we have these concrete cases recorded. In our knowledge Prisma works just fine and returns the data that was requested, just using a different method.

But yes, we are working on fixing this so that Prisma uses the queries that people expect.

**Comment by seromenho** [comment]
@janpio There's a related issue here: https://github.com/prisma/prisma/issues/13306
Created a reproducible: https://github.com/seromenho/prisma-join-performance
A failing job: https://github.com/seromenho/prisma-join-performance/actions/runs/5181630196/jobs/9337397224

Although it's not the same it's related. My reproduction makes it fail and this issue is only about being slow...

**Comment by janpio** [comment]
That issue has been closed half a year ago. Is that still happening in recent Prisma versions? If so, happy to re-open @seromenho.

**Comment by seromenho** [comment]
@janpio I've updated Prisma to the latest version (v4.15.0) on the reproducible repo and that issue (https://github.com/prisma/prisma/issues/13306) is not happening anymore. We don't need to reopen it 👍 Thanks

Although we can see that using a proper join is at least 6 times faster than not. And this rate increase with the amount of data. It helps with this issue. 

https://github.com/seromenho/prisma-join-performance/actions/runs/5198505887/jobs/9374710869
```
    ✓ findMany (4776 ms)
    ✓ findMany and take only some (1027 ms)
    ✓ Raw query with LEFT JOIN (relatively fast ~1s) (761 ms)
```

**Comment by janpio** [comment]
Thanks for confirming that there is no error anymore. We are aware of the potential performance implications in some cases and will take that into account. (We are actively looking into this, so I expect us to have news soon.)

**Comment by mhesham93** [comment]
> Thanks for confirming that there is no error anymore. We are aware of the potential performance implications in some cases and will take that into account. (We are actively looking into this, so I expect us to have news soon.)

Hey Janpio, appreciate all your work on this however the issue still exists for me. I have a query that runs a report with a date range, it works fine if I fetch a month worth of data but anything more than that reproduces the error. 

Thank you

**Comment by janpio** [comment]
Please open a new issue and provide all the information the bug issue template asks for @mhesham93 - we are happy to look into such problems if we have some information to do so. (Your reproduction will also be a good case to test against when we implement native JOINs for such queries.)

**Comment by Scalahansolo** [comment]
Just adding to the noise here... This is a massive issue with Prisma. I understand the reasoning for not using native joins with the "multiple db" connections on the horizon, but the lack of using even the most basic joins in Prisma causes massive performance hits that basically make Prisma unusable at scale. 

Basically with every other ORM out there, you would expect nested where to operate on joins rather than inner selects. At scale, inner selects cause crippling performance problems, especially if those inner selects don't hit indexes. Prisma has best in class developer experience, but in this case, it ends up being a massive issue. Developers end up writing queries that seems super reasonable at the surface, but when deployed at scale cause huge isssues. 

Take this query for example...
```ts
const workspaceMemberByUser = await this.prisma.workspaceMember.findMany({
  where: {
    userId: workspaceMember,
    workspace: {
      name: {
        not: 'Personal',
      },
      teamId: null,
    },
  },
  include: {
    workspace: true,
  },
})
```

Seems totally reasonable, right? Well the issue is, the nested where on `workspace` does massive full table scans on a scaled prod database because the it doesn't rely on the join from the `where: userId: workspaceMember`. The nested where literally has to scan ever row in the workspace table. 

I love Prisma and the power it provides to developers to move fast, but this is a basic function of a SQL database and literally why you choose relational databases, and Prisma just throws that benefit aside. It's hard to recommend to co-workers to continue using Prisma when this land mine is lurking on literally every query we write.

**Comment by janpio** [comment]
I can quote myself from what I wrote not long ago above:

> We are actively looking into this, so I expect us to have news soon.

Or from the message before that:

> But yes, we are working on fixing this so that Prisma uses the queries that people expect.

I hope that is explicit enough. 
We never give ETAs on _when_ something is shipped, so I this is as concrete as it gets. 

Does that satisfy your question?

---

> I understand the reasoning for not using native joins with the "multiple db" connections on the horizon

I have no idea where this comes from, but this is not a relevant consideration for us right now. Maybe might have been 4 years ago, but multi database is not a priority right now, but making Prisma as good as possible is. Hence we focus on solving the problems our users actually have right now. (And even if multi database needed to happen soon, we would just use different queries for that to optimize the much more common use case we are talking about here.)


> Seems totally reasonable, right? Well the issue is, the nested where on `workspace` does massive full table scans on a scaled prod database because the it doesn't rely on the join from the `where: userId: workspaceMember`. The nested where literally has to scan ever row in the workspace table.

This example you provide by the way is exactly what I/we keep asking for. We need concrete cases like this where the current behavior clearly shows performance disadvantages that we can double check against any new implementation. If you have the time, please split that out into a bug report issue and provide some more information about the generated query you observe, and the problem that is concretely causing.

**Comment by inveve** [comment]
We are leaving and migrating away from Prisma for this issue. AN issue open for over 2 years, wow.

**Comment by andrictham** [comment]
I think, if you’ve acknowledged that this is a huge pain point for users, and you’ve committed to improving it, and you want users to give feedback as to test cases for benchmarking performance, then:

The issue here is that [the docs](https://www.prisma.io/docs/concepts/components/prisma-schema/relations) doesn’t communicate the current behavior which doesn’t use database-level joins for relation queries, but instead emulates it. 

Relevant pages in the docs could link to this issue, and an issue-reporting form which solicits those cases you’ve asked for. Along with a guide on temporary workarounds (like using raw queries or pairing it with a 3rd-party generator).

This might save folks a world of pain from adopting the library for join-heavy queries, and then later finding out about this issue the hard way.

I used to be a PM, and I know that it’s impossible to estimate when WIP items will land in production, but at the very least, communicate that this is next in line to land and provide temporary solutions. I think that would help manage expectations better and give people the confidence to continue using Prisma despite its flaws (knowing that the team has acknowledged it, that it will be improved shortly, and that there are ways to work around it in the meantime).

**Comment by andrictham** [comment]
> PS: Small aside: There are already queries where Prisma uses JOINs. It's just not in all of the ones where it would be possible and would make sense.

While this issue is being resolved, could there be documentation which explains which types of queries emulate joins, and which uses database-level joins? 

Knowing how the library will behave can help developers plan their queries and avoid surprises.

**Comment by Scalahansolo** [comment]
I meant no hostility in either of my above comments. I'm just trying to understand the current state of this internally better, and more specifically what does "we are working on it" mean? Does the team understand the scope of the issue / does addressing the root of this issue align with where the team wants to take Prisma?

Like has been pointed out, this issue is roughly 2 years old. I don't fully understand how reporting new issues / examples is helpful or necessary here. Even the most basic queries with Prisma don't leverage joins, but if it would be helpful I could give even more examples.

**Comment by inveve** [comment]
They are **not** working on this issue and suggest we use "Raw Queries"... 

Prisma does not work for companies at scale

**Comment by janpio** [comment]
@Scalahansolo:

> I meant no hostility in either of my above comments.

Thank you, I appreciate you calling that out.

A lot is lost in written communication, but I hope you understand how difficult this is for me here as well. So please also take my answer with the assumption there is no hostility.

> and more specifically what does "we are working on it" mean?

I am not really sure what you are asking me. 

But let me give it a shot again and explicitly spell it out what is happening:

We are currently writing code for Prisma that will use JOINs instead of emulating that by doing separate queries and then combining in memory. The code is not ready, so you can not use it yet. But it will be, and then we will update this issue with a comment that some Prisma Client queries now use JOINs instead of emulating them. We will keep doing so until there are no queries left where we think we can do that (and there are many where, what Prisma allows you to do in one query, is impossible to reflect with database native JOINs), and then we will close this issue.

(Again, I have no ETA when any of this will be made available to you. If someone complains in a week or month or whenever that it has been that long since I posted this update, it will just show again why no matter what we do, someone will find it wrong.)

> I don't fully understand how reporting new issues / examples is helpful or necessary here. Even the most basic queries with Prisma don't leverage joins, but if it would be helpful I could give even more examples.

We know the SQL queries Prisma generates. 
We know many of those do not use JOINs. 
We do **not** know what _concrete_ problems these queries create _for you_ (beyond the abstract "they should use JOINs").

So there are a lot of comments and complaints in our GitHub issues, but very few concrete reproducible examples of real performance problems caused by Prisma's current approach to JOINs (which, see above, we are already working on changing!) that we can use to ensure that we do not accidentally make performance worse in some cases. Prisma generates many, many complex queries - and just replacing everything with JOINs where it is _technically_ possible has the real risk that we accidentally make some use cases worse.

**Comment by janpio** [comment]
@andrictham Thanks for the suggestion. 

Most users are happy, as Prisma gives them the data they ask for and does it with acceptable (to them) performance. Optimizing the documentation to point out where features are missing, or where _some_ users think we should do things differently, does not scale for us. With the user numbers Prisma has, and how we acknowledge and handle most sensible requests, instead of just closing them - that would double the size of the docs.

What you suggest, to ask for examples of performance problems, I am doing here in this issue already (see my explanation in the comment before why this is useful for us). Unfortunately, I have not seen many issues with concrete performance problems having been opened. So we are still left with having to come up potential performance problems ourselves, instead of being able to use the ones that seem to exist out there (based on the interest this issue has).

**Comment by inveve** [comment]
@janpio You guys are by nature making what could be a single query with a join numerous queries due to the lack of joins. Abstraction is better than a concrete example. Instead of a single request there are multiple requests. Data is pulled separately and then combined. This increases latency when a database is large. It really makes no sense for a relational db and is leading to noticeable performance issues in our software. Performance was acceptable for the first 6 months, but as our db got larger, Prisma hsd become the bottleneck to performance.

**Comment by Scalahansolo** [comment]
@janpio Thank you for the follow up. I'll keep a close eye on upcoming releases for improvements.

I've got two follow up questions / comments:

1. In the upcoming Prisma Releases, would it be possible to call out improvements on this front in release notes? Would help those of us who are running into perf issues to test queries where we would expect to see improvements.
2. The biggest offenders for this issue right now, from my experience are queries with nested where that don't hit indexes.

I've run into some kind of strange edge cases lately as I've been more familiar with this issue. I'll try to keep in mind to update here when I run across queries I believe should be using Joins, but aren't.

**Comment by moonman239** [comment]
I learned about Prisma yesterday, thinking it would make my SQL querying simpler. I like how Prisma has a MEDIAN function - that's what drove me to use it to begin with.
Today, though, I have a table to join, and I am concerned that Prisma doesn't support joins.
Does anyone know of a decent alternative to Prisma?

**Comment by Scalahansolo** [comment]
> I learned about Prisma yesterday, thinking it would make my SQL querying simpler. I like how Prisma has a MEDIAN function - that's what drove me to use it to begin with. Today, though, I have a table to join, and I am concerned that Prisma doesn't support joins. Does anyone know of a decent alternative to Prisma?

Let's leave this thread about the issue at hand with Prisma. There are plenty of other options out there for JS / TS ORMs or Typed SQL libs. A simple google search will get ya what you need.

**Comment by tonybruess** [comment]
> I learned about Prisma yesterday, thinking it would make my SQL querying simpler. I like how Prisma has a MEDIAN function - that's what drove me to use it to begin with. Today, though, I have a table to join, and I am concerned that Prisma doesn't support joins. Does anyone know of a decent alternative to Prisma?

I had great luck using Kysely. It even has a Prisma generator.

**Comment by inveve** [comment]
> I learned about Prisma yesterday, thinking it would make my SQL querying simpler. I like how Prisma has a MEDIAN function - that's what drove me to use it to begin with. Today, though, I have a table to join, and I am concerned that Prisma doesn't support joins. Does anyone know of a decent alternative to Prisma?

drizzle #1

**Comment by janpio** [comment]
@moonman239 Prisma supports queries for relations, so you can get the data you probably want. It just currently uses separate queries to do so (one for the main results, and then one for the data of the relation), which sometimes leads to worse performance over a query that would have used a `JOIN`. 

As I said multiple times in this thread, we are aware that this is not great and are working on fixing this. If you can live with the not optimal performance for a while, Prisma can still be a great choice.

**Comment by janpio** [comment]
@tufailra97 There is progress. Last release included some performance-related improvements that are topically pretty close to the JOINs topic: https://github.com/prisma/prisma/releases/tag/5.1.0 It will still take a bit to make it into a release, Prisma is complex.

**Comment by Petrucheqa** [comment]
> We are leaving and migrating away from Prisma for this issue. AN issue open for over 2 years, wow.

Your are making business with this project, feel free to contribute if you want to help.

**Comment by nikooiko** [comment]
> @andrictham Thanks for the suggestion.
> 
> Most users are happy, as Prisma gives them the data they ask for and does it with acceptable (to them) performance. Optimizing the documentation to point out where features are missing, or where _some_ users think we should do things differently, does not scale for us. With the user numbers Prisma has, and how we acknowledge and handle most sensible requests, instead of just closing them - that would double the size of the docs.
> 
> What you suggest, to ask for examples of performance problems, I am doing here in this issue already (see my explanation in the comment before why this is useful for us). Unfortunately, I have not seen many issues with concrete performance problems having been opened. So we are still left with having to come up potential performance problems ourselves, instead of being able to use the ones that seem to exist out there (based on the interest this issue has).

With the purpose/hope of assisting on that end, I'll share an example that I think showcases performance issues, mostly related to the nature these extra queries are executed (ofc this would be eliminated with joins):
If the relations are increasing (e.g. running a query with 4-5 or even more related tables, or even worse with more nested relations with intermediate levels) the initial query can easily generate 8-9-10 or even more queries. Although this can be ok in some cases, in others can cause major issues:
1. The latency caused by these roundtrips increases drastically as the callee must wait for many subsequent queries to (I) join the query queue, (ii) be sent to DB, (iii) for their response to arrive and (iv) the response to be aggregated and combined to the final result. **This can and will take some time** => probably with very large Qs can even reach the default (or any other reasonable) query timeout
2. The Prisma query queue instead of having to deal with 1 query needs to serve many more. If this query runs for an endpoint that scales, it can easily congest the queue, delaying not only the response of that query (1) but also the other queries and business logic.

What I'm trying to say here is that the initial issue (of not using joins or even subqueries) not only can affect the DB level operations but also the performance and reliability of the entire system. It can easily lead to many queries waiting in the queues (especially if these queues aren't properly configured to handle this) and in the end many timeouts that render the system unusable. This isn't farfetched, believe I saw this in action and the only resolution is to use alternate paths like raw queries or query builder.

**Comment by tobold** [comment]
We are struggling with this at the moment. We are trying to get data for many items from another related item in another table. If we were writing the SQL for this, we would be using a join. Prisma is instead passing tens of thousands of variables into a query which is causing the database to throw an error `too many bind variables in prepared statement, expected maximum of 32767`.

Surely it should be noted in the documentation that Prisma doesn't support queries of more than 32767 joined items? In enterprise software this is a massive limitation.

**Comment by janpio** [comment]
You are obviously hitting a bug @tobold. Can you please open a new issue and describe your query? Thanks.

**Comment by mochetts** [comment]
@janpio with the most of respects. It's not a bug.. it's a whole feature missing.

**Comment by janpio** [comment]
Then please open an issue where you describe more about `too many bind variables in prepared statement, expected maximum of 32767` instead of talking about it in a feature request to use JOINs (which we are already working on, as I have acknowledged multiple times above). Then we can understand what query exactly understand is causing it, and make sure that when we implement this feature that bug goes away. Thank you.

**Comment by adam-arold** [comment]
> You are obviously hitting a bug @tobold. Can you please open a new issue and describe your query? Thanks.

I'm wondering if you intended that as a joke (bad case) or if you really don't understand how SQL works (worst case) and how Prisma **is supposed** to work as literally everybody here expects apart from you.

**Comment by janpio** [comment]
@adam-arold I am pretty sure I understand how Prisma is supposed to work ❤️ 

We have explicit logic in place ("chunking") that should split SQL queries that create too many bind variables in a prepared statement into multiple SQL queries, so that this error message does not happen. If that does not work, that is a bug and should be reported as such.

Note that this logic, and my response, in no way means that I do not agree that Prisma could use `JOIN` queries in more cases. As mentioned multiple times before in this thread, we are working on that and have been releasing improvements to that consistently in the last few releases already.

But even if the queries that are currently failing for @tobold and @mochetts would not fail the same way after we also implemented `JOIN`s for them, we are still interested in understanding the problems in our chunking implementation as it is used in other places where `JOIN`s can not be used.

**Comment by AurelianSpodarec** [comment]
So theres no way to use prisma and JOIN? 

So that means instead of making say 1query, I would need to make 5queries because JOIN does not exist?

**Comment by nathanchapman** [comment]
> So theres no way to use prisma and JOIN?
> 
> So that means instead of making say 1query, I would need to make 5queries because JOIN does not exist?

https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries

**Comment by AurelianSpodarec** [comment]
> > So theres no way to use prisma and JOIN?
> > So that means instead of making say 1query, I would need to make 5queries because JOIN does not exist?
> 
> https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries

So but that doesn't utilise the JOIN right. Thus its not as performant right.

**Comment by borodiliz** [comment]
> > > So theres no way to use prisma and JOIN?
> 
> > > So that means instead of making say 1query, I would need to make 5queries because JOIN does not exist?
> 
> > 
> 
> > https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries
> 
> 
> 
> So but that doesn't utilise the JOIN right. Thus its not as performant right. 

https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access

**Comment by AurelianSpodarec** [comment]
> > > > So theres no way to use prisma and JOIN?
> > 
> > 
> > > > So that means instead of making say 1query, I would need to make 5queries because JOIN does not exist?
> > 
> > 
> > > 
> > 
> > 
> > > https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries
> > 
> > 
> > So but that doesn't utilise the JOIN right. Thus its not as performant right.
> 
> https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access

But that doesn't provide assistance when you're writing raw queries like some other ORMs

**Comment by Scalahansolo** [comment]
Y'all, please read the thread before starting to just add more comments like this. This is a known issue that Prisma is actively working on improve. Jan has stated that multiple times on this thread. We all know there are still certain sets of queries that are non optimal. Prisma also made some big strides in this regard with the 5.4 release. 

The last couple of releases seem to have been focused on Serverless / database driver compatibility. I think we are all hoping that 5.7 / 5.8 have more improvements that relate to this thread and other performance type issues.

**Comment by Weakky** [comment]
Hey folks,

We’re excited to let you know that we’ve just added support for `JOIN`s to resolve relations in [[Prisma 5.7.0](https://github.com/prisma/prisma/releases/tag/5.7.0)](https://github.com/prisma/prisma/releases/tag/5.7.0) for PostgreSQL and CockroachDB, under the `relationJoins` preview feature.

Once you enable that feature, `JOIN`s will be used to resolve one-to-one, one-to-many and many-to-many relations. Most of the queries listed on [[this documentation page](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-reads)](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-reads) will essentially now use joins. This means that, when possible, we now resolve your query in a single round-trip as opposed to the N round-trips we used to make for each relation.

I say *when possible*, because we fallback to the old query-based mechanism when you:

- Use `distinct`
- Use `cursor`
- Use `_count` in your selection set

This is the **very first step** to bring `JOIN`s to Prisma for querying relations. We plan on removing the current limitations and adding support for other databases in future releases.

Changing the way Prisma resolves relations is not a trivial change and is likely not going to work 100% on day one, so please bear with us as we continuously improve the feature, both in terms of bugs and performance improvements on the SQL queries we generate.

If you are curious, we have provided a bunch of Prisma query examples together with their associated SQL query output in our [[latest release notes](https://github.com/prisma/prisma/releases/tag/5.7.0)](https://github.com/prisma/prisma/releases/tag/5.7.0).

Finally, we have a dedicated [[preview feature feedback issue here](https://github.com/prisma/prisma/discussions/22288)](https://github.com/prisma/prisma/discussions/22288) if you have any thoughts about it.

Thank you all for your patience!

**Comment by moonman239** [comment]
Thank you. I am sure this will be a welcome update.

**Comment by CyanoFresh** [comment]
does the resulting query looks ugly or is just for me? (json, nested select from select - 4-level when we need only 2)

**Comment by janpio** [comment]
They are indeed a bit ugly, but fortunately, what matters is that they return the correct data with the lowest latency 🙈 

If you find any queries where that is _not_ true (correctness or performance), please do [report them in our feedback discussion](https://github.com/prisma/prisma/discussions/22288) - that is _exactly_ why we do these as a preview feature.

The JSON and "levels" confused me a bit as well in the beginning, but I was convinced by my colleagues (and the benchmarks they showed me 🤷) that these are actually faster and more efficient. With normal plain joins a lot of data is duplicated (in a left join, the left side is repeated for each entry on the right side), which has to be transmitted from database to database client, only to be thrown away and deduplicated as we want to return a object with sub-objects anyway. 

(And again, if that is not true in your cause or causes any trouble, let us know.)

Ah, and we have an outstanding issue to make the aliases and general query structure a bit "prettier". But that will come later when we have enough feedback that they actually do their job properly.

**Comment by michaelhays** [comment]
Super excited for this, nice work! Any idea when MySQL will get support for this as well?

(I know y'all don't like providing estimates for understandable reasons, just wondering if that's a relatively straightforward follow-on, or something that may take 6+ months)

**Comment by controversial** [comment]
> The JSON and "levels" confused me a bit as well in the beginning, but I was convinced by my colleagues (and the benchmarks they showed me 🤷) that these are actually faster and more efficient. With normal plain joins a lot of data is duplicated (in a left join, the left side is repeated for each entry on the right side), which has to be transmitted from database to database client, only to be thrown away and deduplicated as we want to return a object with sub-objects anyway.

I would be very interested to read a blog post on this subject in the future!

Congrats on the release of such a big/complex feature :)

**Comment by jekh** [comment]
> The JSON and "levels" confused me a bit as well in the beginning, but I was convinced by my colleagues (and the benchmarks they showed me 🤷) that these are actually faster and more efficient. With normal plain joins a lot of data is duplicated (in a left join, the left side is repeated for each entry on the right side), which has to be transmitted from database to database client, only to be thrown away and deduplicated as we want to return a object with sub-objects anyway.

Is this true for 1:1 relations too? There shouldn't be any duplicated data from the join (so no bandwidth penalty).

Are there known negative performance impacts from the lateral join? I would speculate (having done no actual benchmarking) that the optimizer is going to be more ...optimized... for regular joins and will probably choose better query plans.

**Comment by ravivarma4003** [comment]
Is joins released for my sql in prisma ?

**Comment by michaelhays** [comment]
> Is joins released for my sql in prisma ?

@ravivarma4003 Not yet, but we're hoping that Prisma 5.10.0 (slated to be released on Feb 20) includes MySQL support for `relationJoins`; see #23010

**Comment by ravivarma4003** [comment]
const assetQuery: { [key: string]: string } = {};
      assetQuery[syncKeyColumn] = syncKeyValue;

      const { asset_number, serial_number } = assetQuery;
      const conditions = [];
      const queryParams: any[] = [organizationId];

      if (asset_number) {
        conditions.push(`a.asset_number = ?`);
        queryParams.push(syncKeyValue); // Use syncKeyValue directly as parameter
      }

      if (serial_number) {
        conditions.push(`a.serial_number = ?`);
        queryParams.push(syncKeyValue); // Use syncKeyValue directly as parameter
      }

      const conditionsString = conditions.length > 0 ? `AND ${conditions.join(' AND ')}` : '';

      const result = await tx.$queryRaw`
  SELECT *
  FROM assets a
  LEFT JOIN mdm_assets m ON a.id = m.asset_id
  WHERE m.asset_id IS NULL
    AND a.organization_id = ${organizationId}
    ${conditionsString}
  LIMIT 1;
`;
This query is not working. Can some one help me with this query..

@michaelhays : Prisma is throwing error 4",
            "message": "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '?\n  LIMIT 1' at line 6

**Comment by ravivarma4003** [comment]
@janpio How to provide a dynamic query in prisma using raw query

**Comment by ravivarma4003** [comment]
async findAssetByFieldsForMdmFlow(
    tx: Prisma.TransactionClient,
    organizationId: IdType,
    assetQuery: AssetQueryDto,
  ): Promise<any> {
    const { asset_number, serial_number } = assetQuery;
    const conditions = [];
    const queryParams: any[] = [organizationId];
    if (asset_number) {
      conditions.push(`a.asset_number = ?`);
      queryParams.push(asset_number);
    }
    if (serial_number) {
      conditions.push(`a.serial_number = ?`);
      queryParams.push(serial_number);
    }
    const conditionsString =
      conditions.length > 0 ? `AND ${conditions.join(' AND ')}` : '';
    const result = await tx.$queryRaw`
    SELECT *
    FROM assets a
    LEFT JOIN mdm_assets m ON a.id = m.asset_id
    WHERE m.asset_id IS NULL
      AND a.organization_id = ${organizationId}
      ${conditionsString}
    LIMIT 1;
  `;
  
  export declare class AssetQueryDto {
    asset_number?: string;
    serial_number?: string;
}

**Comment by aqrln** [comment]
@mrroll 

> Do these cases still cause the relation to fall back to `query`?

`_count` is now supported with `join` (both on PostgreSQL/CockroachDB and on MySQL).

`distinct` is supported on PostgreSQL (except cases where `distinct` fields are incompatible with `orderBy`) but not on CockroachDB and MySQL yet. Full support for `distinct` should be shipped in the next release, a PR is open, it just didn't make it into this release.

`cursor` still isn't supported yet but it's in the plan.

**Comment by janpio** [comment]
Right now we have not plans to add this to MariaDB as it unfortunately neither supports lateral joins, the method we use for PostgreSQL, nor correlated subqueries, the method we use for MySQL. We would need to try a completely different approach for MariaDB :/

**Comment by janpio** [comment]
3 new issues tracking the demand for the addition of this functionality for SQL Server, SQLite and MariaDB:
https://github.com/prisma/prisma/issues/23346
https://github.com/prisma/prisma/issues/23347
https://github.com/prisma/prisma/issues/23348

**Comment by exelimpichment** [comment]
> > Is joins released for my sql in prisma ?
> 
> Now, yes: https://github.com/prisma/prisma/releases/tag/5.10.0 🥳

this does not work with planetScale, right? 
as far as I am concerned they run mySQL 8.0 or I am wrong?

**Comment by wmadden** [comment]
I'm closing this issue as a duplicate since as noted [above](https://github.com/prisma/prisma/issues/5184#issuecomment-1977323638), we have separate issues to track adding support for `relationJoins` to SQL Server, SQLite and MariaDB.
